internal/internal_task_handlers.go

```diff
@@ -1622,28 +1622,64 @@ type cadenceInvoker struct {
 	cancelHandler         func()
 	heartBeatTimeoutInSec int32       // The heart beat interval configured for this activity.
 	hbBatchEndTimer       *time.Timer // Whether we started a batch of operations that need to be reported in the cycle. This gets started on a user call.
-	lastDetailsToReport   *[]byte
+	detailsToReport       *[]byte     // Details to be reported in the next reporting interval.
+	lastDetailsReported   *[]byte     // Details that were reported in the last reporting interval.
 	closeCh               chan struct{}
 	workerStopChannel     <-chan struct{}
 }
 
-func (i *cadenceInvoker) Heartbeat(details []byte, skipBatching bool) error {
+func (i *cadenceInvoker) Heartbeat(details []byte) error {
 	i.Lock()
 	defer i.Unlock()
 
-	if i.hbBatchEndTimer != nil && !skipBatching {
+	_, err := i.internalHeartBeat(details)
+	return err
+}
+
+func (i *cadenceInvoker) BackgroundHeartbeat() error {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.hbBatchEndTimer != nil {
+		if i.detailsToReport == nil {
+			i.detailsToReport = i.lastDetailsReported
+		}
+
+		return nil
+	}
+
+	var details []byte
+	if i.detailsToReport != nil {
+		details = *i.detailsToReport
+	} else if i.lastDetailsReported != nil {
+		details = *i.lastDetailsReported
+	}
+
+	return i.heartbeatAndScheduleNextRun(details)
+}
+
+func (i *cadenceInvoker) BatchHeartbeat(details []byte) error {
+	i.Lock()
+	defer i.Unlock()
+
+	if i.hbBatchEndTimer != nil {
 		// If we have started batching window, keep track of last reported progress.
-		i.lastDetailsToReport = &details
+		i.detailsToReport = &details
 		return nil
 	}
 
+	return i.heartbeatAndScheduleNextRun(details)
+}
+
+func (i *cadenceInvoker) heartbeatAndScheduleNextRun(details []byte) error {
 	isActivityCancelled, err := i.internalHeartBeat(details)
 
 	// If the activity is cancelled, the activity can ignore the cancellation and do its work
 	// and complete. Our cancellation is co-operative, so we will try to heartbeat.
-	if (err == nil || isActivityCancelled) && !skipBatching {
+	if err == nil || isActivityCancelled {
```

Comment:
I don't understand this...why we still need to do the work if activity is already canceled?